crates/rbuilder/src/utils/provider_factory_reopen.rs

```diff
@@ -129,3 +152,168 @@ pub fn check_provider_factory_health<DB: Database>(
+        provider.sealed_header(number)
+    }
+
+    fn sealed_headers_while(
+        &self,
+        range: impl RangeBounds<BlockNumber>,
+        predicate: impl FnMut(&SealedHeader) -> bool,
+    ) -> ProviderResult<Vec<SealedHeader>> {
+        let provider = self
+            .check_consistency_and_reopen_if_needed()
+            .map_err(|e| ProviderError::Database(DatabaseError::Other(e.to_string())))?;
+        provider.sealed_headers_while(range, predicate)
+    }
+}
+
+impl<DB: Database + Clone> BlockHashReader for ProviderFactoryReopener<DB> {
+    fn block_hash(&self, number: BlockNumber) -> ProviderResult<Option<B256>> {
+        let provider = self
+            .check_consistency_and_reopen_if_needed()
+            .map_err(|e| ProviderError::Database(DatabaseError::Other(e.to_string())))?;
+        provider.block_hash(number)
+    }
+
+    fn canonical_hashes_range(
+        &self,
+        start: BlockNumber,
+        end: BlockNumber,
+    ) -> ProviderResult<Vec<B256>> {
+        let provider = self
+            .check_consistency_and_reopen_if_needed()
+            .map_err(|e| ProviderError::Database(DatabaseError::Other(e.to_string())))?;
+        provider.canonical_hashes_range(start, end)
+    }
+}
+
+impl<DB: Database + Clone> BlockNumReader for ProviderFactoryReopener<DB> {
+    fn chain_info(&self) -> ProviderResult<ChainInfo> {
+        self.provider_factory_unchecked().chain_info()
+    }
+
+    fn best_block_number(&self) -> ProviderResult<BlockNumber> {
+        self.provider_factory_unchecked().best_block_number()
+    }
+
+    fn last_block_number(&self) -> ProviderResult<BlockNumber> {
+        self.provider_factory_unchecked().last_block_number()
+    }
+
+    fn block_number(&self, hash: B256) -> ProviderResult<Option<BlockNumber>> {
+        let provider = self
+            .check_consistency_and_reopen_if_needed()
+            .map_err(|e| ProviderError::Database(DatabaseError::Other(e.to_string())))?;
+        provider.block_number(hash)
+    }
+}
+
+impl<DB: Database + Clone> BlockIdReader for ProviderFactoryReopener<DB> {
+    fn pending_block_num_hash(&self) -> ProviderResult<Option<BlockNumHash>> {
+        unimplemented!("This method is not supported by ProviderFactoryReopener. Please consider using a BlockchainProvider.");
+    }
+
+    fn safe_block_num_hash(&self) -> ProviderResult<Option<BlockNumHash>> {
+        unimplemented!("This method is not supported by ProviderFactoryReopener. Please consider using a BlockchainProvider.");
+    }
+
+    fn finalized_block_num_hash(&self) -> ProviderResult<Option<BlockNumHash>> {
+        unimplemented!("This method is not supported by ProviderFactoryReopener. Please consider using a BlockchainProvider.");
+    }
+}
+
+impl<DB: Database + Clone> StateProviderFactory for ProviderFactoryReopener<DB> {
+    fn latest(&self) -> ProviderResult<StateProviderBox> {
+        self.provider_factory_unchecked().latest()
```

Comment:
The check should be added here too.

Reply:
we use history_by_block hash IIRC so its not critical

Reply:
I added it to all methods